aiohttp/web_request.py

```diff
@@ -510,10 +511,42 @@ def text(self):
         return bytes_body.decode(encoding)
 
     @asyncio.coroutine
-    def json(self, *, loads=json.loads):
-        """Return BODY as JSON."""
+    def json(self, *, loads=json.loads, allowed_types=None):
+        """ Return BODY as JSON. If the `Content-Type` is not in allowed_types,
+        then the request is rejected.
+        """
+        content_type = self.content_type
+        (mime_type, subtype, suffix, parameters) = parse_mimetype(content_type)
+        content_type = mime_type + '/' + subtype
+        if suffix != '':
+            content_type += '+' + suffix
+
+        charset = parameters.get('charset', None)
+
+        if isinstance(allowed_types, str):
+            if allowed_types == 'json':
+                if content_type != 'application/json':
+                    if not content_type.startswith('application') or \
+                       not content_type.endswith('+json'):
+                        raise HTTPNotAcceptable()
+            elif allowed_types != content_type:
+                raise HTTPNotAcceptable()
+        elif isinstance(allowed_types, set):
+            if content_type not in allowed_types:
+                raise HTTPNotAcceptable()
+        elif isinstance(allowed_types, tuple):
+            if content_type not in allowed_types:
+                raise HTTPNotAcceptable()
+        elif isinstance(allowed_types, list):
+            if content_type not in allowed_types:
+                raise HTTPNotAcceptable()
+        elif allowed_types is None:  # do not check the content_type
+            pass
+        else:  # for other type
+            raise TypeError("Unsupported type for allowed `Content-Type`")
+
         body = yield from self.text()
-        return loads(body)
+        return loads(body, encoding=charset)
```

Comment:
The other arguments have the same meaning as in load(), except encoding which is ignored and deprecated.